src/java.base/share/classes/java/lang/IO.java

```diff
@@ -0,0 +1,191 @@
+        System.out.println(obj);
+    }
+
+    /**
+     * Writes a line separator to the standard output.
+     *
+     * <p> The effect is as if {@link java.io.PrintStream#println() println()}
+     * had been called on {@code System.out}.
+     */
+    public static void println() {
+        System.out.println();
+    }
+
+    /**
+     * Writes a string representation of the specified object to the
+     * standard output.
+     *
+     * <p> The effect is as if {@link java.io.PrintStream#print(Object) print(obj)}
+     * had been called on {@code System.out}.
+     *
+     * @param obj the object to print, may be {@code null}
+     */
+    public static void print(Object obj) {
+        System.out.print(obj);
+        System.out.flush();
+    }
+
+    /**
+     * Reads a single line of text from the standard input.
+     * <p>
+     * If necessary, this method first creates internal objects that decode
+     * the bytes read from the standard input into characters. These objects
+     * are retained for subsequent reuse by this method.
+     * <p>
+     * One line is read from the decoded input as if by
+     * {@link java.io.BufferedReader#readLine() BufferedReader.readLine()}
+     * and then the result is returned.
+     *
+     * @return a string containing the line read from the standard input, not
+     * including any line separator characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if an I/O error occurs
+     */
+    public static String readln() {
+        try {
+            return reader().readLine();
+        } catch (IOException ioe) {
+            throw new IOError(ioe);
+        }
+    }
+
+    /**
+     * Writes a prompt and then reads a line of input.
+     *
+     * <p>
+     * Writes a prompt as if by calling {@code print}, and then reads a single
+     * line of text as if by calling {@link readln readln()}.
+     *
+     * @param prompt the prompt string, may be {@code null}
+     *
+     * @return a string containing the line read from the standard input, not
+     * including any line separator characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if an I/O error occurs
+     */
+    public static String readln(String prompt) {
+        print(prompt);
+        return readln();
+    }
+
+    /**
+     * The BufferedReader used by readln(). Initialized under a class lock by
+     * the reader() method.
+     */
+    static BufferedReader br;
+
+    /**
+     * On the first call, creates an InputStreamReader to decode characters from
+     * System.in, wraps it in a BufferedReader, and returns the BufferedReader.
+     * These objects are cached and returned by subsequent calls.
+     *
+     * @return the internal BufferedReader instance
+     */
+    static synchronized BufferedReader reader() {
```

Comment:
Is the lock only inteneded for initialization, i.e. should it apply for simply getting the reader?

Reply:
I believe there needs to be some barriers (at least `volatile`) even for getting the already initialized reader. And then, given the code here, it is probably fine and less error prone to simply use `synchronized` that something more complex.

But, given we'll hopefully have `@StableValue` soon, maybe we could use that? What do you think @stuart-marks?

Thanks!

Reply:
It's even more complex than `volatile`. If a thread comes along and reads a null here, it still needs to take a lock (or something) to prevent more than one thread from creating the BufferedReader and storing it back into the field. This leads down the path of double-checked locking, initialize-on-demand holder, etc.

It's not clear to me this is in such a critical path that it warrants adding any of this stuff. StableValue probably has the right semantics. However, it's not even in the JDK yet (though it is Targeted to 25 as of this moment). But I don't think we want to add a dependency on a Preview API.